// @eximchain/dappsmith/dist/dappSrc/components/reusable/ui/NavLeft/ButtonSmall.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
require("./Button.css");
exports.ButtonStyles = {
    primary: 'primary',
    secondary: 'secondary',
    tertiary: 'tertiary',
    outline: 'outline',
    destructive: 'destructive',
    disabled: 'disabled',
    hovered: 'hovered',
    pressed: 'pressed',
    idle: 'idle',
    close: 'close',
    cancel: 'cancel'
};
exports.SizeMap = {
    small: 'small',
    large: 'large',
    medium: 'medium'
};
function prefix(baseClass, style) {
    return baseClass + (style ? "-" + style : "");
}
exports.prefix = prefix;
var noop = function () { };
var getClasses = function (props) {
    var customClass = props.customClass, className = props.className;
    var styles = {};
    //Apply classes based on props
    Object.keys(exports.ButtonStyles).filter(function (key) {
        styles[prefix(customClass, key)] = props[key];
    });
    return classnames_1.default(customClass, prefix(props.customClass, exports.SizeMap.small), styles, className);
};
exports.ButtonSmall = function (props) {
    var onClick = props.onClick, disabled = props.disabled, children = props.children;
    var classes = getClasses(props);
    return (React.createElement("button", { className: classes, onClick: !disabled ? onClick : noop, disabled: disabled }, children ? children : React.createElement("span", null, "Button")));
};
exports.ButtonSmall.defaultProps = {
    disabled: false,
    customClass: 'btn',
    hovered: false,
};
exports.default = exports.ButtonSmall;
