// @eximchain/dappsmith/dist/dappSrc/components/reusable/ui/NavLeft/ButtonLink.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
require("./ButtonLink.css");
function prefix(baseClass, style) {
    return baseClass + (style ? "-" + style : "");
}
exports.prefix = prefix;
exports.ButtonLinkStyles = {
    plain: 'plain',
    link: 'link',
    address: 'address',
    disabled: 'disabled',
    selected: 'selected',
    hovered: 'hovered',
    pressed: 'pressed',
    idle: 'idle'
};
var noop = function (e) { return e.stopPropagation(); };
var getClasses = function (props) {
    var customClass = props.customClass, className = props.className;
    var styles = {};
    Object.keys(exports.ButtonLinkStyles).filter(function (key) {
        styles[prefix(customClass, key)] = props[key];
    });
    return classnames_1.default(customClass, styles, className);
};
exports.ButtonLink = function (props) {
    var onClick = props.onClick, disabled = props.disabled, children = props.children, href = props.href;
    var classes = getClasses(props);
    return (React.createElement("a", { className: classes, onClick: !disabled ? onClick : noop, href: href }, children ? children : React.createElement("span", null, "Button Link")));
};
exports.ButtonLink.defaultProps = {
    disabled: false,
    href: '#',
    customClass: 'anchor',
    link: false,
    plain: false,
    address: false,
    pressed: false,
    hovered: false,
    idle: false
};
exports.default = exports.ButtonLink;
