// @eximchain/dappsmith/dist/dappSrc/components/reusable/ui/NavLeft/ButtonCancel.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
require("./Button.css");
exports.ButtonStyles = {
    primary: 'primary',
    secondary: 'secondary',
    tertiary: 'tertiary',
    outline: 'outline',
    destructive: 'destructive',
    disabled: 'disabled',
    hovered: 'hovered',
    pressed: 'pressed',
    idle: 'idle',
    close: 'close',
    cancel: 'cancel'
};
exports.SizeMap = {
    small: 'small',
    large: 'large',
    medium: 'medium'
};
function prefix(baseClass, style) {
    return baseClass + (style ? "-" + style : "");
}
exports.prefix = prefix;
var noop = function () { };
var ButtonCancel = /** @class */ (function (_super) {
    __extends(ButtonCancel, _super);
    function ButtonCancel(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            hovered: !!_this.props.hovered,
            className: _this.getClasses()
        };
        _this.onMouseOver = _this.onMouseOver.bind(_this);
        _this.onMouseOut = _this.onMouseOut.bind(_this);
        return _this;
    }
    ButtonCancel.prototype.getClasses = function () {
        var _a = this.props, customClass = _a.customClass, className = _a.className;
        var styles = {};
        if (this.state) {
            styles[prefix(customClass, exports.ButtonStyles.hovered)] = this.state.hovered;
        }
        this.setState({
            className: classnames_1.default(customClass, styles, className)
        });
        return classnames_1.default(customClass, styles, className);
    };
    ButtonCancel.prototype.onMouseOver = function () {
        this.setState({
            hovered: true
        }, this.getClasses);
    };
    ButtonCancel.prototype.onMouseOut = function () {
        this.setState({
            hovered: false
        }, this.getClasses);
    };
    ButtonCancel.prototype.render = function () {
        var _a = this.props, onClick = _a.onClick, disabled = _a.disabled, children = _a.children;
        return (React.createElement("button", { className: this.state.className, onClick: !disabled ? onClick : noop, disabled: disabled, onMouseOver: this.onMouseOver, onMouseOut: this.onMouseOut }, children ? children : React.createElement("span", null, "Cancel")));
    };
    ButtonCancel.defaultProps = {
        disabled: false,
        customClass: 'btn-cancel',
        hovered: false
    };
    return ButtonCancel;
}(React.Component));
exports.ButtonCancel = ButtonCancel;
exports.default = ButtonCancel;
