// @eximchain/dappsmith/dist/dappSrc/components/reusable/ui/NavLeft/Label.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
require("./Label.css");
exports.LabelColors = {
    green: 'green',
    purple: 'purple',
    yellow: 'yellow',
    blue: 'blue',
    red: 'red',
    darkNeutral: 'darkneutral',
    lightNeutral: 'lightneutral',
};
exports.LabelSize = {
    big: 'big',
    small: 'small'
};
exports.LabelStatus = {
    sent: 'sent',
    pending: 'pending',
    received: 'received'
};
function prefix(baseClass, style) {
    return baseClass + (style ? "-" + style : "");
}
exports.prefix = prefix;
var getClasses = function (props) {
    var customClass = props.customClass, size = props.size, className = props.className, color = props.color, status = props.status, outline = props.outline;
    var styles = {};
    if (status && exports.LabelStatus[status]) {
        styles[prefix(customClass, exports.LabelStatus[status])] = true;
        return classnames_1.default(customClass, styles, className);
    }
    styles[prefix(customClass, 'outline')] = outline;
    if (size && exports.LabelSize[size])
        styles[prefix(customClass, exports.LabelSize[size])] = true;
    if (color && exports.LabelColors[color])
        styles[prefix(customClass, exports.LabelColors[color])] = true;
    return classnames_1.default(customClass, styles, className);
};
exports.Label = function (props) {
    var children = props.children;
    var classes = getClasses(props);
    return (React.createElement("label", { className: classes }, children ? children : React.createElement("span", null)));
};
exports.Label.defaultProps = {
    outline: false,
    size: 'small',
    customClass: 'label'
};
exports.default = exports.Label;
