// @eximchain/dappsmith/dist/dappSrc/components/reusable/fields/shared.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
exports.inputUpdater = function (onChange, options) {
    if (options === void 0) { options = {}; }
    var withError = options.withError, clean = options.clean;
    return function (e) {
        var val = e.currentTarget.value;
        if (withError)
            withError('');
        if (clean)
            val = clean(val);
        onChange(val);
    };
};
exports.inputValidator = function (validator, withError, errorMsg) {
    if (errorMsg === void 0) { errorMsg = "Invalid value."; }
    return function (e) {
        if (!validator(e.target.value)) {
            withError(errorMsg);
        }
    };
};
exports.displayLabel = function (name, displayName) {
    return displayName ? react_1.default.createElement("label", { style: {
            display: 'block',
            paddingTop: '1em',
            paddingBottom: '0.25em',
            fontWeight: 'bold'
        }, htmlFor: name }, displayName) : null;
};
