// @eximchain/dappsmith/dist/dappSrc/components/reusable/ui/Modal/Modal.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var Box_1 = __importDefault(require("./../Box"));
var Button_1 = __importDefault(require("./../Button"));
var Divider_1 = __importDefault(require("./../Divider"));
var Icon_1 = __importDefault(require("./../Icon"));
var Text_1 = __importDefault(require("./../Text"));
var StopScrollBehavior_1 = __importDefault(require("./../utils/StopScrollBehavior"));
var TrapFocusBehavior_1 = __importDefault(require("./../utils/TrapFocusBehavior"));
var OutsideEventBehavior_1 = __importDefault(require("./../utils/OutsideEventBehavior"));
var StyledModal_1 = __importDefault(require("./StyledModal"));
var SIZE_WIDTH_MAP = {
    xs: 314,
    sm: 414,
    md: 544,
    lg: 804
};
var ESCAPE_KEY_CODE = 27;
var Backdrop = function (_a) {
    var children = _a.children;
    return (React.createElement(React.Fragment, null,
        React.createElement("div", { className: "Backdrop" }),
        children));
};
var Modal = /** @class */ (function (_super) {
    __extends(Modal, _super);
    function Modal() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.handleOutsideClick = function () {
            _this.props.onDismiss();
        };
        _this.handleCloseClick = function () {
            _this.props.onDismiss();
        };
        _this.handleKeyUp = function (event) {
            if (event.keyCode === ESCAPE_KEY_CODE) {
                _this.props.onDismiss();
            }
        };
        _this.renderHeader = function () {
            var _a = _this.props, heading = _a.heading, role = _a.role, showCloseButton = _a.showCloseButton;
            var justifyContent = role === "popup" ? "center" : "start";
            if (role === "dialog" || (role === "popup" && showCloseButton)) {
                return (React.createElement(Box_1.default, { display: "flex", justifyContent: "center", paddingY: 6, paddingX: 12 },
                    React.createElement(Text_1.default, { bold: true }, heading)));
            }
            else {
                return (React.createElement(Box_1.default, { display: "flex", padding: 6, justifyContent: justifyContent },
                    React.createElement(Text_1.default, { bold: true }, heading)));
            }
        };
        _this.renderCloseButton = function () {
            var _a = _this.props, accessibilityCloseLabel = _a.accessibilityCloseLabel, role = _a.role, showCloseButton = _a.showCloseButton;
            if (role === "dialog" || (role === "popup" && showCloseButton)) {
                return (React.createElement(Box_1.default, { padding: 3, position: "absolute", top: true, right: true },
                    React.createElement(Button_1.default, { accessibilityLabel: accessibilityCloseLabel, className: "close", onClick: _this.handleCloseClick },
                        React.createElement(Icon_1.default, { icon: "close" }))));
            }
            return null;
        };
        return _this;
    }
    Modal.prototype.componentDidMount = function () {
        window.addEventListener("keyup", this.handleKeyUp);
    };
    Modal.prototype.componentWillUnmount = function () {
        window.removeEventListener("keyup", this.handleKeyUp);
    };
    Modal.prototype.render = function () {
        var _a = this.props, accessibilityModalLabel = _a.accessibilityModalLabel, children = _a.children, footer = _a.footer, _b = _a.role, role = _b === void 0 ? "dialog" : _b, _c = _a.size, size = _c === void 0 ? "sm" : _c;
        var width = SIZE_WIDTH_MAP[size];
        return (React.createElement(StopScrollBehavior_1.default, null,
            React.createElement(TrapFocusBehavior_1.default, null,
                React.createElement(StyledModal_1.default, { "aria-label": accessibilityModalLabel, className: "container", role: role },
                    React.createElement(Backdrop, null,
                        React.createElement(OutsideEventBehavior_1.default, { onClick: this.handleOutsideClick },
                            React.createElement("div", { className: "wrapper", tabIndex: -1, style: { width: width } },
                                React.createElement(Box_1.default, { flex: "grow", direction: "column", display: "flex", position: "relative", width: "100%" },
                                    React.createElement(Box_1.default, { fit: true },
                                        this.renderHeader(),
                                        this.renderCloseButton(),
                                        role === "dialog" && React.createElement(Divider_1.default, { type: "secondary" })),
                                    React.createElement(Box_1.default, { flex: "grow", overflow: "auto", position: "relative", paddingX: 6, paddingY: 3 },
                                        React.createElement(Text_1.default, null, children)),
                                    React.createElement(Box_1.default, { fit: true }, footer && (React.createElement(Box_1.default, null,
                                        role === "dialog" && React.createElement(Divider_1.default, { type: "secondary" }),
                                        React.createElement(Box_1.default, { padding: 6 },
                                            React.createElement(Text_1.default, null, footer)))))))))))));
    };
    return Modal;
}(React.Component));
exports.default = Modal;
