// @eximchain/dappsmith/dist/dappSrc/components/reusable/ui/Button/Button.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
var StyledButton_1 = __importDefault(require("./StyledButton"));
var noop = function () { };
//Button as stateless react component as we dont't require any state here.
exports.Button = function (props) {
    var onClick = props.onClick, accessibilityExpanded = props.accessibilityExpanded, accessibilityHaspopup = props.accessibilityHaspopup, accessibilityLabel = props.accessibilityLabel, block = props.block, children = props.children, className = props.className, disabled = props.disabled, loader = props.loader, size = props.size, style = props.style, theme = props.theme, type = props.type;
    var classes = classnames_1.default("btn", style, style == "close" || style == "complete" ? "" : theme, size, className, {
        block: block
    });
    var renderContent = function () {
        if (loader)
            return React.createElement(React.Fragment, null, "Loading..");
        else if (children)
            return children;
        else
            return React.createElement(React.Fragment, null, "Button");
    };
    return (React.createElement(StyledButton_1.default, { onClick: disabled || loader ? noop : onClick, "aria-expanded": accessibilityExpanded, "aria-haspopup": accessibilityHaspopup, "aria-label": accessibilityLabel, className: classes, disabled: disabled, type: type }, renderContent()));
};
exports.Button.defaultProps = {
    disabled: false,
    loader: false,
    size: "large",
    style: "standard",
    theme: "cta",
    type: "button",
    block: false
};
exports.Button.displayName = "Button";
exports.default = exports.Button;
